src/state/Data/actions.js

```diff
@@ -8,24 +9,113 @@ import request from '../_common/request';
+ * @return {Number}
  */
-const isSpatialDataEmpty = (spatialData) => {
-    if(spatialData && !_.isEmpty(spatialData) ) {
-        const sd = Object.values(spatialData);
-        return !sd.some(data => data?.data && !_.isEmpty(data.data));
+const getPageSize = (state) => {
+    const localConfig = Select.app.getCompleteLocalConfiguration(state);
+    const PAGE_SIZE = localConfig.requestPageSize || configDefaults.requestPageSize;
+    return PAGE_SIZE;
+}
+
+const getRestRelationsPages = (attributeRelationsCount, spatialRelationsCount, PAGE_SIZE) => {
+    // What is higer to load? attributeRelations or spatialRelations
+    const maxCountValue = Math.max(attributeRelationsCount, spatialRelationsCount);
+    if(maxCountValue === 0) {
+        return 0;
     } else {
-        return true;
+        const remainingPageCount = Math.ceil((maxCountValue - PAGE_SIZE) / PAGE_SIZE);
+        return remainingPageCount;
+    }
+}
+
+/**
+ * Helper function. Usually second step in requesting data.
+ * Calculate if relations requests are missing based on attributeRelationsCount and spatialRelationsCount.
+ * Each relations request loads one next tile from spatialFilter.
+ * Rest tiles are loaded without relatiions.
+ * @param {bool} loadGeometry Whether response should contain geometry
+ * @param {Object} spatialFilter Spatial defined filter of level and its tiles
+ * @param {string?} styleKey UUID
+ * @param {Array?} order
+ * @param {Object} mergedSpatialFilter Filler object contains modifiers and layerTemplateKey or areaTreeLevelKey.
+ * @param {Object} mergedAttributeFilter Filler object contains modifiers, layerTemplateKey or areaTreeLevelKey and styleKey.
+ * @param {Number} attributeRelationsCount Count of known attribute relations. Used for determinate further requests.
+ * @param {Number} spatialRelationsCount Count of known spatial relations. Used for determinate further requests.
+ * @param {Array} preloadedSpatialDataSources SpatialDataSources loaded by previous request.
+ * @return {function} Return promise. 
+ */
+function loadMissingRelationsAndData(loadGeometry, spatialFilter, styleKey, order, mergedSpatialFilter, mergedAttributeFilter, attributeRelationsCount, spatialRelationsCount, preloadedSpatialDataSources = []) {
+    return (dispatch, getState) => {
+        const PAGE_SIZE = getPageSize(getState());
+
+        // FIXME - add attributeFilter support
+        // attributeFilter is null at the moment
+        const attributeFilter = null;
+        const dataSourceKeys = null;
+        const featureKeys = null;
+        const promises = [];
+
+        // load remaining relations pages
+        const remainingRelationsPageCount = getRestRelationsPages(attributeRelationsCount, spatialRelationsCount, PAGE_SIZE);
+        let tilesPagination = 0;
+        const loadRelations = true;
+        for (let i = 0; i < remainingRelationsPageCount; i++) {
+            const relations = {
+                offset: (i + 1) * PAGE_SIZE,
```

Comment:
Maybe cycle from i = 1, if we only end up using i+1.

Reply:
https://github.com/gisat-panther/ptr-state/pull/56/commits/8d782033c9b74d6479a478a71df3caa456aa74ca changes the number of cycles. Should probably be <= now?